_posts/2022-03-07-using-private-image-repositories.markdown

```diff
@@ -0,0 +1,201 @@
+    Name:                  nginx-deployment
+    Phase:                 Created
+...
+Events:
+  Type     Reason  Age                  From                       Message
+  ----     ------  ----                 ----                       -------
+  Warning  Failed  4m10s (x2 over 16m)  edgedeployment-controller  error playing YAML file: initializing source docker://quay.io/jdzon/nginx:1.14.2: reading manifest 1.14.2 in quay.io/jdzon/nginx: unauthorized: access to the requested resource is not authorized
+```
+
+#### Creating auth file secret
+To use private container image, user needs to obtain credentials for their image repository - contents of the auth file generated when one executes
+```bash
+podman login quay.io
+```
+
+Usually that would be `$HOME/.docker/config.json` or `${XDG_RUNTIME_DIR}/containers/auth.json` file.
+
+When you have the auth file, you need to create a kubernetes secret containing it under `.dockerconfigjson` key:
+```bash
+kubectl create secret generic pull-secret \
+  --type=kubernetes.io/dockerconfigjson \
+  --from-file=.dockerconfigjson=${XDG_RUNTIME_DIR}/containers/auth.json
+```
+
+The created secret will look similar to the following one:
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: pull-secret
+data:
+  .dockerconfigjson: ewoJImF1dGhzIjogewoJCSJxdWF5LmlvIjogewoJCQkiYXV0aCI6ICJabTl2SUdKaGNpQm1iMjhnWW1GeUNnPT0iCgkJfQoJfQp9Cg== 
+type: kubernetes.io/dockerconfigjson
+```
+
+#### Deploying the workload
+
+To deploy workloads with a private image stored in the repository, you need to instruct Flotta operator to use the 
+secret with valid credentials for the repository hosting that image. You do that by referencing it in the EdgeDeployment 
+specification, under `imageRegistries`:
+
+```yaml
+apiVersion: management.project-flotta.io/v1alpha1
+kind: EdgeDeployment
+metadata:
+  name: nginx-deployment
+spec:
+  ...
+  imageRegistries:
+    secretRef:
+      name: pull-secret
+  ...
+```
+
+
+After you update your `EdgeDeployment`:
+
+```shell
+kubectl apply -f -<<EOF
+apiVersion: management.project-flotta.io/v1alpha1
+kind: EdgeDeployment
+metadata:
+  name: nginx-deployment
+spec:
+  type: pod
+  deviceSelector:
+    matchLabels:
+      dc: home
+  imageRegistries:
+    secretRef:
+      name: pull-secret
+  pod:
+    spec:
+      containers:
+        - name: nginx
+          image: quay.io/jdzon/nginx:1.14.2
+          ports:
+            - containerPort: 80
+              hostPort: 9090
+EOF
+```
+After the next heartbeat received from the device, you will see the phase of the deployment to be `Running`:
+```shell
+$ kubectl describe edgedevice
+Name:         fedora-54b9dd17-9971-11ec-bcee-7085c256610d
+Namespace:    default
+Labels:       dc=home
+              device.cpu-architecture=x86_64
+...
+Status:
+  Deployments:
+    Last Transition Time:  2022-03-07T08:37:12Z
+    Name:                  nginx-deployment
+    Phase:                 Running
+...
+```
+
```

Comment:
do you think there is a need to show the nginx url responding to curl on port 9090 ?

Reply:
No. I think that would be a good thing to show in a blog post describing how to deploy software to edge devices in general (continuation of https://project-flotta.github.io/flotta.github.io/flotta/2022/02/03/getting-started-project-flotta.html).